### ISSUE TITLE
PKG Add jax

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide-recipes#187]
Looking at the list of dependencies [jax](https://github.com/google/jax) should be easier to build than torch or tensorflow pyodide/pyodide-recipes#175, and it would be useful to include it.

**Comment by anutosh491** [comment]
Hey @femtomc @josephrocca 

Are the $500 bounties still alive ? as mentioned [here](https://github.com/pyodide/pyodide-recipes/issues/187#issuecomment-3096483941) and [here](https://github.com/pyodide/pyodide-recipes/issues/187#issuecomment-3096484019) 

JAX and XLA are now running entirely in the browser. Here's a quickstart demo.

https://github.com/user-attachments/assets/142e49b8-e145-4088-b75f-e151e4a73d41

My implementation targets JupyterLite, xeus-python and emscripten-forge rather than Pyodide. I maintain emscripten-forge and intend to package the work there.

I can provide working & reproducible JAX & XLA in Jupyterlite links soon if that's good enough to claim the bounty. Cheers !

**Comment by josephrocca** [comment]
@anutosh491 👀 Can you share a minimal self-contained html snippet so I can test? The snippet can fetch external js/wasm of course. I'm also curious about limitations - i.e. presumably there was a lot of code you had to remove/ignore in the port. Would be great to have that documented.

I don't need this anymore (four years since my original attempt 😅) but still quite happy to respect the original bounty because jax in the browser is cool¹. 

@ryanking13 I did [already send the bounty](https://github.com/pyodide/pyodide-recipes/issues/187#issuecomment-3096484019) to the pyodide opencollective, but completely understandable if that has already been used for something else - will gladly send another $500 (either to the opencollective for subsequent payout or directly to @anutosh491's github sponsors page).

¹ BTW, there's now [JAX.js](https://github.com/ekzhang/jax-js) by @ekzhang which I'm sure is several OOMs faster and more practical for standard JAX-in-the-browser usage, but it's still neat to have the emscripten port. Might be some future opportunity to bring GPU inference to Pyodide/xeus-python using JAX.js behind the scenes, since I assume this emscripten port doesn't support GPU-side stuff.

**Comment by ryanking13** [comment]
> @ryanking13 I did https://github.com/pyodide/pyodide-recipes/issues/187#issuecomment-3096484019 to the pyodide opencollective, but completely understandable if that has already been used for something else - will gladly send another $500 (either to the opencollective for subsequent payout or directly to @anutosh491's github sponsors page).

Sure we can pay $500 from our open collective funding. I am also curious what kind of features are available and not available in JAX compiled to WASM

**Comment by anutosh491** [comment]
Thanks @josephrocca and @ryanking13 — here it is 😃

<img width="500" height="600" alt="Image" src="https://github.com/user-attachments/assets/7b028a87-580c-41ec-98ba-6e7b7e4281f3" />

- Minimal zero-click [HTML demo](https://anutosh21.github.io/jax-wasm-demo/) . The source is [here](https://github.com/anutosh491/jax-wasm-demo/blob/main/site/index.html)
- [Interactive notebook](https://anutosh21.github.io/jax-wasm-demo/lab/index.html?path=jax_quickstart_wasm.ipynb) with jax.jit and jax.grad

i) The first visit downloads the browser-native Python and XLA runtime, so cold startup depends on network speed. The runtime is cached in the visitor’s browser, making subsequent visits considerably faster. So forgive me for the demos being slower for the first time. 
ii) I randomly saw memory out of bounds errors couple times. These demos are using my local builds and I don't expect it to show once we host a proper (avoiding any hacks) recipe on emscripten-forge. So forgive me if you might have to press reload here and there.


Limitations : I would say these are expected ones based on the patches I've made now.

- CPU execution only; GPU, TPU, and WebGPU backends are not included.
- wasm32 with 32-bit pointers and the browser's corresponding memory limits. Emscripten-forge would target wasm64 soon !
- Single-threaded execution. Distributed, multi-host, MPI, and Gloo paths are excluded. 

I've only tried scalar JIT and automatic-differentiation examples for now, the complete JAX test suite has not been run but I'm decently confident we'll do well with vector and tensor operations too. I could even try a standard notebooks from upstream jax.

Please give it a try. Hopefully you should be able to run my links (kinda rushed through the implementation for them 😬 ). Let me know if something doesn't work for y'all 

That being said I've accepted payments through both methods in the past. Offcourse the gh sponsorship page is easier I guess cause it's one shot isn't it ? and I remember opening expenses on lfortran's opencollective when I used to contribute there (possibly would have to do the same on pyodide's open collective !?) . Whatever y'all would be comfortable with ... I'll go ahead with the same. Thanks again ;)

**Comment by anutosh491** [comment]
> I've only tried scalar JIT and automatic-differentiation examples for now, the complete JAX test suite has not been run but I'm decently confident we'll do well with vector and tensor operations too. I could even try a standard notebooks from upstream jax.

I tried some stuff from [quickstart.ipynb](https://github.com/jax-ml/jax/blob/main/docs/pallas/quickstart.ipynb)

https://github.com/user-attachments/assets/9be8360d-b27f-4477-b6d0-13b962f4284f

And I think It works well.

**Comment by hoodmane** [comment]
To claim the bounty, I'd assume we want a jax recipe that allows us to reproduce the build locally or in CI.

**Comment by josephrocca** [comment]
> Minimal zero-click [HTML demo](https://anutosh21.github.io/jax-wasm-demo/) . The source is [here](https://github.com/anutosh491/jax-wasm-demo/blob/main/site/index.html)

@anutosh491 Any chance you could provide a version that's as minimal as this: https://github.com/josephrocca/ftfy-pyodide/blob/main/mod.js ? I.e. just loading the whl with pyodide and executing some JAX code. The HTML source you linked is indeed a small/minimal file, but it achieves that by just embedding the JupyterLite web app as an iframe 😅

**Comment by anutosh491** [comment]
> To claim the bounty, I'd assume we want a jax recipe that allows us to reproduce the build locally or in CI.

Requiring a reproducible recipe or CI build does make sense.

One point I’d like to clarify with @josephrocca  is whether the bounty is about achieving reproducible JAX JIT execution in the browsers/Jupyterlite powered by wasm or it is specifically tied to Pyodide packaging ?

My implementation runs the original JAX/XLA stack through JupyterLite, xeus-python and emscripten-forge. Since I maintain emscripten-forge ([repo](https://github.com/emscripten-forge/recipes) & [docs](https://emscripten-forge.org/))—and this builds directly on my recent [llvmlite and Numba JIT](https://notebook.link/blog/numba-in-the-browser/) work there—that would naturally be the first home for the recipe !

Python in the browser is broader than any single distribution, so would an emscripten-forge recipe, CI-reproducible build and public browser demo satisfy the bounty, or is a merged Pyodide recipe an explicit requirement?

I’d like to align on the expected deliverable before packaging it.

**Comment by anutosh491** [comment]
> Any chance you could provide a version that's as minimal as this

Sure I shall look into it tomorrow and write back (kinda late as per Indian Standard Time).

**Comment by hoodmane** [comment]
I think a reproducible emscripten-forge recipe would be fine too. But I suppose since it's @josephrocca's money it ought to be up to him.

**Comment by ryanking13** [comment]
Yeah, as a Pyodide maintainer, it would be great if we can have working jax in Pyodide and if possible, upstream it to jax repository as well, but since it is josephrocca's money, we can pay if he is happy with it.

**Comment by anutosh491** [comment]
> Any chance you could provide a version that's as minimal as this: https://github.com/josephrocca/ftfy-pyodide/blob/main/mod.js ? I.e. just loading the whl with pyodide and executing some JAX code. 

Thanks @josephrocca !

I’ve now reduced the demo to a standalone browser page—there is no `xeus-python` kernel, `JupyterLite` application, `notebook`, `iframe`, or remote Python server involved.

https://github.com/user-attachments/assets/f1281df3-7366-46f6-8509-4f817b24e6b6



- The page first initializes `CPython` compiled to WebAssembly. Once the runtime is ready, the user can edit and run ordinary Python code using jax.jit (The initial visit downloads and initializes the WebAssembly runtime; the browser can cache these static assets, and subsequent executions in the same tab reuse the already-running Python runtime)
- I've first imported jax and then run some jax jit code. 
- The visible `[XLA/Wasm]` logs show the runtime compilation path: XLA emits a WebAssembly object, invokes an in-process WebAssembly build of LLD, links a shared Wasm module, loads it with dlopen, resolves the compiled symbol, and executes it locally in the browser. 

These logs are explicitly added by me to show there's no cheating involved here and everything's in wasm running in wasm ;)

Hopefully this is enough for claiming the bounty. Here's the minimal link to try out : https://anutosh21.github.io/jax-wasm-demo/

Also for @hoodmane @ryanking13 I shall add the recipe to emscripten-forge soon and share it over once done. Hope that works :)

**Comment by josephrocca** [comment]
@anutosh491 Nice work! I'll let @hoodmane and @ryanking13 handle stuff from here. Take the list below with a grain of salt since it's a coding agent's work/output, but it does seem like there's some necessary/commonly-used stuff missing, and also some not 100% necessary (but nice-to-have stuff) if you can manage it.

---

## ✅ Working (verified against real JAX 0.11.2 — outputs match)

- `jax.jit` (incl. `static_argnames`, compile cache)
- `grad`, `value_and_grad`, `jvp`, `vjp`, `jacfwd`, `hessian`
- `vmap`, `pmap` (1 device)
- Pytrees (`tree_map`, flatten/unflatten)
- `lax.cond`, `lax.fori_loop`
- Core `jnp` ops: matmul, sum, cumsum, sort/argsort, concat, clip, where, take, einsum, top_k, segment_sum, slicing, `.at[]` (contiguous)
- dtypes: float64 (true precision), float32, float16, bfloat16, complex64, ints, bool, type promotion
- Convolutions (`lax.conv_general_dilated`, `jax.scipy.signal.convolve`)
- FFT (`fft`, `rfft`, `ifft`)
- `jax.nn`: relu, silu, gelu (both modes), softmax, log_softmax, sigmoid, elu, swish, selu, one_hot
- `jax.scipy.special`: erf, erfc, i0e, ndtri, log_ndtr
- `jax.scipy.stats`: norm, poisson
- `custom_jvp` / `custom_vjp`, `remat`, checkpoint policies
- `make_jaxpr`, `eval_shape`, AOT (`lower()` / `compile()` / `as_text()`)
- MLIR/StableHLO Python bindings
- `pure_callback`, `jax.debug.print` (works under jit)
- `Mesh` / `NamedSharding` / `with_sharding_constraint`
- `jax.example_libraries.optimizers` (adam)
- `jax_enable_x64`

## ❌ Not working, but should be (all fixable at build level — not WASM limitations)

- **`jax.random.*` (all PRNG impls)** — aborts the entire runtime: `Thread tf_foreach creation via pthread_create() failed`. The XLA CPU PRNG kernel needs a serial fallback in single-threaded wasm. Kills the whole JAX 101 random-numbers chapter.
- **`lax.scan`** (and `associative_scan`, `lax.map`) — fails to link its JIT'd module: `wasm-ld: error: relocation R_WASM_MEMORY_ADDR_LEB cannot be used against symbol '.L__constant_Nxf32'; recompile with -fPIC`. The codegen emits **non-PIC objects**; this also hits `jnp.arange`/`iota` at certain shapes, `jax.image.resize`, strided `.at[::2].add()`, and `einsum`/`jacfwd` at some shapes — it looks nondeterministic but is shape-dependent. Highest-impact single fix.
- **`lax.while_loop`** — hard wasm trap: `memory access out of bounds`, kills the kernel.
- **`jax.nn.leaky_relu`** — same hard trap.
- **`jax.scipy.special.digamma` / `gammaln` / `betainc`** — same hard trap.
- **LAPACK-backed linalg: `eigh`, `svd`, `qr`, `cholesky`, `solve`, `inv`, `lstsq`** (both `jnp.linalg` and `jax.scipy.linalg`, incl. `expm`, `lu_factor`) — `NameError: name '_lapack' is not defined`. The LAPACK extension wasn't built in; Emscripten/Pyodide already proves LAPACK-in-wasm is viable.
- **`jax.experimental.sparse` (BCOO)** — unusable: `scipy` isn't packaged.
- **`jax.export` serialization** — `flatbuffers` isn't packaged.
- **`jax.profiler`** — top-level attribute missing.
- **Threading** — single-threaded only (no SharedArrayBuffer, no COOP/COEP headers on GH Pages). Acceptable fallback if thread-pool kernels get serial paths — but currently they abort (see `jax.random`).

Not expected / fine to omit: GPU/TPU backends, multi-device pmap, distributed.

**Comment by anutosh491** [comment]
Woah that's a long list !! Thank you for it.

Offcourse I didn't test my work against all of jax's test suite cause I just got it working yesterday !

But I shall be trying to address a good chunk of these things once I host the recipe on emscripten-forge. I'll reference your comment for things to be fixed.

My take

> lax.scan (and associative_scan, lax.map) — fails to link its JIT'd module: wasm-ld: error: relocation R_WASM_MEMORY_ADDR_LEB cannot be used against symbol '.L__constant_Nxf32'; recompile with -fPIC. 

- I've fixed errors like these before and should be fixable. 
- We have lapack, scipy and flatbuffers on emscripten-forge so those should be doable too.
- I'm challenged by the `memory access out of bounds` errors really.

So I guess I do have your vote on the bounty you announced !? (which should also motivate me to solve the above bugs as I'm doing this in my free time).

**Comment by josephrocca** [comment]
> So I guess I do have your vote on the bounty you announced !? (which should also motivate me to solve the above bugs as I'm doing this in my free time).

Yep, I think you have the bounty locked for sure - it'd just be a matter of solving some of those final issues, and satisfying @hoodmane and @ryanking13 on the recipe side. I think it's reasonable to ship this without threading support, btw, in case that turns out to be a bigger headache. But the commonly-used features that currently crash due to lack of threading support (like `jax.random.*`) would need a single-threaded fallback, as the agent output above mentions.

While I'm here: I had a separate agent independently do the same kind of analysis - I figured it's worth attaching this one too because there seems to either be some statefulness or randomness to the crashes:

> Crash behavior is **nondeterministic**: the same program can run fine one time and crash the next (e.g. `logsumexp` passed 3/6 runs; an MLP loop completed bit-identically once then crashed 5/5 attempts)

Hopefully that behavior goes away once you've solved the other issues.

Second agent's full summary below:

---

*Tested build: the runtime shipped by anutosh21.github.io/jax-wasm-demo (CPython 3.13.1, NumPy 2.4.6, JAX/JAXLib 0.11.2.dev, real XLA CPU compiler). Reference: real JAX 0.11.1 on CPU. ~515 isolated probe items + all 8 official JAX starter notebooks were run under both.*

## ✅ Working (verified, outputs match real JAX)

- **Core transforms:** `jax.jit` (array/static args, `donate_argnums`, caching, `.lower()/.compile()`), `grad`, `value_and_grad`, `jvp`, `vjp`, `linearize`, `hessian`, `jacrev`, `custom_jvp`, `custom_vjp`, `checkpoint`/remat, `vmap` (on runtime/numpy-typed arrays), `pmap` (single device), `shard_map` (single device), `with_sharding_constraint`, `Array.sharding`

- **Control flow:** `lax.cond`, `lax.switch`, `lax.while_loop`, `lax.fori_loop`, `lax.scan` (on runtime arrays), `jnp.where`, `jnp.clip`

- **`jax.numpy`** (when inputs are numpy-provided arrays, not jnp constants): elementwise math, broadcasting, indexing incl. `.at[...]`, `matmul`/`dot`/`einsum`/`tensordot`, `fft`, `norm`, `trace`, x64 float64

- **Pytrees & interop:** `jax.tree_util` (tree_map, flatten, transpose, register_pytree_node), np↔jnp conversion, `device_put`, `block_until_ready`

- **`jax.scipy` (pure-Python paths):** `special` (gammaln, digamma, erf/erfc, erfinv, gammainc, betainc, i0/i1, logsumexp, expit, logit, ndtr/ndtri, zeta, fresnel, …), `stats` (norm, uniform, beta, gamma, poisson, bernoulli, binom, t, chi2, …), `optimize.minimize` (BFGS), `interpolate`, `ndimage`, `signal`

- **Misc:** `jax.nn` activations (softmax/relu/gelu/one_hot/sigmoid) + `zeros/ones/constant` initializers, `jax.image.resize`, `jax.experimental.ode.odeint`, `jax.debug.print/callback`, `jax.stages` (lower/compile)

- **Correctness:** a full `jit`+`grad`+`vmap` MLP training loop produced **bit-identical loss values to real JAX** (in runs that didn't crash); a 35-item value suite matched to ≤1 ULP.

## ❌ Non-working but should ideally work (all CPU-runnable)

- **`jax.random` — every generation function crashes the whole runtime** (`normal`, `uniform`, `randint`, `bernoulli`, `categorical`, `choice`, `permutation`, `gumbel`, `exponential`, `poisson`, `gamma`, `beta`, `dirichlet`, `t`, `pareto`, `weibull_min`, `multivariate_normal`, `truncated_normal`, `orthogonal`, `bits`, `split`, `fold_in`). Root cause: XLA's threefry/thread-pool path calls `pthread_create()`, which fails in single-threaded wasm, and XLA **aborts instead of falling back**. Only `PRNGKey`/`key_data` work. This alone breaks the official quickstart notebook in its first cell.
- **All LAPACK linear algebra** — `NameError: name '_lapack' is not defined`: `jnp.linalg.svd/eigh/eigvalsh/cholesky/qr/lu/solve/inv/pinv/lstsq/det/slogdet/matrix_rank/cond/expm/sqrtm/schur/hessenberg/solve_triangular` and the same under `jax.scipy.linalg`. OpenBLAS was built but excluded from the public bundle (per the project's LIMITATIONS.md), and the missing module isn't handled gracefully.
- **`jax.experimental.sparse` / `jax.scipy.sparse` — cannot even be imported** (`jax/experimental/sparse/linalg.py` does `import scipy.sparse`; `scipy` isn't in the bundle).
- **`jnp.arange`, `jnp.eye`, `jnp.linspace`, and any op whose inputs are jnp constants** — wasm-ld link error (`relocation R_WASM_MEMORY_ADDR_LEB ... recompile with -fPIC`). This also breaks `vmap`/`lax.scan`/`lax.map`/`lax.reduce`/`lax.cumsum`/`slice`/`gather`/`reshape`/`jacfwd`/`jacrev` whenever they see constant arrays — i.e. most tutorial-style code.
- **Python `float`/`jnp.float32` scalar arguments to `jit`** — crash the runtime ("memory access out of bounds"), flaky. `jit(f)(3.0)` is extremely common in educational code. (`jit(f)(3)` ints and 0-d *arrays* `np.array(3.0)`/`jnp.array(3.0)` mostly work.)
- **Reductions & sort/search ops — crash (nondeterministic):** `jnp.sum` on large arrays, `median`, `percentile`/`quantile`, `sort`/`argsort`/`partition`, `searchsorted`/`digitize`, `unique`, `nonzero`/`flatnonzero`, `histogram`, `nanmean`/`nansum`, `lax.rng_uniform`, `lax.argmax/argmin`, `lax.sort`, `lgamma`/`igamma` in some shapes.
- **`jax.nn.initializers`** (`normal`, `uniform`, `xavier_uniform`, `he_normal`, `lecun_normal`, `variance_scaling`) — crash because they call `jax.random`; `nn.initializers.orthogonal` fails on `_lapack`; `nn.standardize` and a few activation paths also crash.
- **`jax.export` / serialization** — not usable.
- **Missing packages in the environment** (`scipy`, `matplotlib`, `rich`, …) — blocks `jax.scipy.sparse` and notebook cells (`vmapped_log_probs` needs `scipy`, `autodiff_remat` needs `rich`).
- **Threading absent and XLA doesn't cope:** `os.cpu_count()==1`, `threading` can't start threads, no `_multiprocessing` — expected for wasm, but XLA still tries to spawn threads and kills the runtime instead of degrading gracefully (the root cause of most items above).

*Correctly absent / not expected: GPU/TPU backends, distributed / multi-host / MPI / Gloo, Triton/Pallas (GPU kernels), jax2tf, profiler traces.*

**Comment by anutosh491** [comment]
Thanks a lot for these pointers. I had literally just tested `jax.jit` on a simple square number operation 😆 

Never knew so much was doable (and so much was not) !

> But the commonly-used features that currently crash due to lack of threading support (like jax.random.*) would need a single-threaded fallback, as the agent output above mentions.

Should be doable !

**Comment by anutosh491** [comment]
Hey all,

A quick follow-up to the JAX/WebAssembly work: the same XLA foundation now has the standard TensorFlow Python package running inside JupyterLite too ;)

https://github.com/user-attachments/assets/b4f2c80e-caef-42fa-845c-a11d37f65f8a

The notebook supports both eager TensorFlow execution and `tf.function(jit_compile=True)`, with XLA compiling tensor programs into WebAssembly inside the browser. I’ve tested automatic differentiation, matrix operations, convolution, attention, and basic Keras inference and training.

This is a useful validation that the JAX result was not a hard-coded demonstration haha —the browser XLA path can support another major frontend and its broader Python ecosystem. Hoping that just makes my claim for the bounty even stronger 😆

Does tensorflow in Jupyterlite interest you @josephrocca ?
